File: cpp-solns/169_majority_element.cpp
#include <vector>
#include <cmath>
#include <unordered_map>

using namespace std;

/* Problem description from LeetCode:

--=========--
Given an array nums of size n, return the majority element.

The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume 
that the majority element always exists in the array.
--=========--

Solution:

--=========--
*/

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int maj = floor(nums.size()/2);

        unordered_map<int, int> counts;

        for (int i = 0; i < nums.size(); i++){
            counts[nums[i]]++;
            if (counts[nums[i]] > maj){
                return nums[i];
            }
        }

        for (const pair<const int, int>& n : counts){
            if (n.second > maj){
                return n.first;
            }
        }

        return 0;
    }
};

File: js-solns/169_majority_element.js
var majorityElement = function(nums) {
    let track = {}
    let res
    const maj = Math.floor(nums.length/2)
    nums.forEach(e => {track[e] = 0})
    nums.forEach(e => {
        track[e]++
        if (track[e] > maj) {
            res = e
            return e
        }
    })
    return res
};